backends-velox/src/main/scala/org/apache/gluten/execution/HashAggregateExecTransformer.scala

```diff
@@ -241,21 +226,21 @@ abstract class HashAggregateExecTransformer(
     }
 
     aggregateFunction match {
-      case hllAdapter: HLLAdapter =>
+      case _ if aggregateFunction.aggBufferAttributes.size > 1 =>
+        generateMergeCompanionNode()
+      case _ =>
         aggregateMode match {
-          case Partial =>
-            // For Partial mode output type is binary.
+          case Partial | PartialMerge =>
             val partialNode = ExpressionBuilder.makeAggregateFunction(
               VeloxAggregateFunctionsBuilder.create(args, aggregateFunction, aggregateMode),
               childrenNodeList,
               modeKeyWord,
               ConverterUtils.getTypeNode(
-                hllAdapter.inputAggBufferAttributes.head.dataType,
-                hllAdapter.inputAggBufferAttributes.head.nullable)
+                aggregateFunction.inputAggBufferAttributes.head.dataType,
```

Comment:
I think the problem here is we assume the `inputAggBufferAttributes.head.dataType` is same with `aggregateFunction.dataType` for partial mode an partialMerge mode. This is not true for HLL. Before we add the special handling for partial mode only and disabled HLL for partialMerge mode.
Thanks.

Reply:
@zhli1142015 Gluten searches agg functions in Velox based on the raw input and final output signatures of the agg function. I think using `inputAggBufferAttributes.head` to construct the `aggFuncNode` for partial agg is inaccurate. It would be more appropriate to follow the original `dataType`.

I searched through the Spark code and found that, for most agg functions, the `dataType` and `inputAggBufferAttributes.head` are the same. However, for decimal avg, the `inputAggBufferAttributes.head` is sum `DecimalType`, and while the final output `resultType` is also `DecimalType`, the precision and scale are different.

Even though Velox does not consider precision and scale when searching agg functions, there might be other agg function types that are completely different. For avg, if `inputAggBufferAttributes` contains count before sum, then the two are entirely inconsistent.

Therefore, I believe using `dataType` is more appropriate as it aligns with most agg functions.
cc @zhztheplayer.

Reply:
I'm little confused about this:

> @zhli1142015 Gluten searches agg functions in Velox based on the raw input and final output signatures of the agg function. I think using `inputAggBufferAttributes.head` to construct the `aggFuncNode` for partial agg is inaccurate. It would be more appropriate to follow the original `dataType`.

Do you mean here I still need to use `dataType` for partial mode even `dataType` and `inputAggBufferAttributes.head` are not same?

Reply:
I think we should use `dataType`, but for HLL, the `dataType` and `inputAggBufferAttributes.head` should also be the same, both being `LongType`, right?

Reply:
Spark HLL `inputAggBufferAttributes` is `Seq[Long]`. We rewrite HLL in Gluten to adapt Velox's function signature, which immediate type is `varbinary`. https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/aggregates/ApproxDistinctAggregate.cpp#L476C33-L476C49.

Reply:
@zhli1142015 I see, I misread some of the code, and your modification is correct.

Reply:
Sure, thanks for your review.